File: in_Python/0917 Reverse Only Letters.py
class Solution(object):
    def reverseOnlyLetters(self, S):
        """
        :type S: str
        :rtype: str
        """
        SS = []
        for i in range(len(S)): SS.append(S[i])
        i = 0
        j = len(S)-1
        while i < j:
            while not S[i].isalpha() and i < len(S)-1:
                i += 1
            while not S[j].isalpha() and j >= 0:
                j -= 1
            if i >= j:
                break
            temp = SS[i]
            SS[i] = SS[j]
            SS[j] = temp
            i += 1
            j -= 1
        res = ""
        for x in SS:
            res += x
        return res
            
        

File: in_Java/0917 Reverse Only Letters.java
class Solution {
    public String reverseOnlyLetters(String S) {
        char[] ch = S.toCharArray();
        int i = 0;
        int j = ch.length-1;
        char t;
        while(i < j){
            while(!isL(ch[i]) && i < j) i++;
            while(!isL(ch[j]) && i < j) j--;
            t = ch[i];
            ch[i] = ch[j];
            ch[j] = t;
            i++;
            j--;
        }
        return String.valueOf(ch);
    }
    private boolean isL(char c){
        if(c >= 'a' && c <= 'z') return true;
        else if(c >= 'A' && c <= 'Z') return true;
        else return false;
    }
}